core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -50,33 +50,98 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
 
   private val shutdownHook = addShutdownHook()
 
+  private abstract class FileAllocationStrategy {
+    def apply(filename: String): File
+
+    protected def getFile(filename: String, storageDirs: Array[File]): File = {
+      require(storageDirs.nonEmpty, "could not find file when the directories are empty")
+
+      // Figure out which local directory it hashes to, and which subdirectory in that
+      val hash = Utils.nonNegativeHash(filename)
+      val dirId = localDirs.indexOf(storageDirs(hash % storageDirs.length))
+      val subDirId = (hash / storageDirs.length) % subDirsPerLocalDir
+
+      // Create the subdirectory if it doesn't already exist
+      val subDir = subDirs(dirId).synchronized {
+        val old = subDirs(dirId)(subDirId)
+        if (old != null) {
+          old
+        } else {
+          val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
+          if (!newDir.exists() && !newDir.mkdir()) {
+            throw new IOException(s"Failed to create local dir in $newDir.")
+          }
+          subDirs(dirId)(subDirId) = newDir
+          newDir
+        }
+      }
+
+      new File(subDir, filename)
+    }
+  }
+
   /** Looks up a file by hashing it into one of our local subdirectories. */
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#getFile().
-  def getFile(filename: String): File = {
-    // Figure out which local directory it hashes to, and which subdirectory in that
-    val hash = Utils.nonNegativeHash(filename)
-    val dirId = hash % localDirs.length
-    val subDirId = (hash / localDirs.length) % subDirsPerLocalDir
-
-    // Create the subdirectory if it doesn't already exist
-    val subDir = subDirs(dirId).synchronized {
-      val old = subDirs(dirId)(subDirId)
-      if (old != null) {
-        old
-      } else {
-        val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
-        if (!newDir.exists() && !newDir.mkdir()) {
-          throw new IOException(s"Failed to create local dir in $newDir.")
-        }
-        subDirs(dirId)(subDirId) = newDir
-        newDir
+  private object hashAllocator extends FileAllocationStrategy {
+    def apply(filename: String): File = getFile(filename, localDirs)
+  }
+
+  /** Looks up a file by hierarchy way in different speed storage devices. */
+  private val hierarchy = conf.getOption("spark.diskStore.hierarchy")
+  private class HierarchyAllocator extends FileAllocationStrategy {
+    case class Level(key: String, threshold: Long, dirs: Array[File])
+    val hsDescs: Array[(String, Long)] =
+      // e.g.: hierarchy = "ram_disk 1GB, ssd 20GB"
+      hierarchy.get.trim.split(",").map {
+        s => val x = s.trim.split(" +")
+          val storage = x(0).toLowerCase
+          val threshold = s.replaceFirst(x(0), "").trim
+          (storage, Utils.byteStringAsBytes(threshold))
```

Comment:
@viirya I update the code here, now if the "spark.diskStore.hierarchy" is not specified correctly, runtime exception will fail the Spark's startup directly. Does it make sense to you?

It is the same to other setting's behavior, like "spark.executor.memory", Spark does not handle its exception also.
```
    if (conf.contains("spark.executor.memory")) {
      val executorMemory = conf.getSizeAsBytes("spark.executor.memory")
      if (executorMemory < MIN_MEMORY_BYTES) {
        throw new IllegalArgumentException(s"Executor memory $executorMemory must be at least " +
          s"$MIN_MEMORY_BYTES. Please increase executor memory using the " +
          s"--executor-memory option or spark.executor.memory in Spark configuration.")
      }
    }
```